### PR TITLE
Fixed #25456 -- Made GenericIPAddressField normalize IPv4 addresses

### PR DESCRIPTION
**Pull request by darjeeling**
Ticket: https://code.djangoproject.com/ticket/25456

### CODE REVIEW

**Comment by hirokiky** [comment]
Checked:

* Nice docstring
* Tests for fields.py
* Tests for utils.ipv4

good to me. another opinion?

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by timgraham** [comment]
If the model field isn't involved, maybe `tests/forms_tests/tests/test_fields.py` is a better place for the test?

**Comment by darjeeling** [comment]
@timgraham oh. Thanks. I think you are right. I'll fix it.

**Comment by timgraham** [comment]
Please uncheck "Patch needs improvement" on the ticket when you update it, thanks.

File: django/utils/ipv4.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+        A normalized IPv4 address, or the same value
+    """
+    octets = ip_str.split(".")
+
+    for index in range(len(octets)):
+        # Remove leading zeroes
+        octets[index] = octets[index].lstrip('0')
+        if not octets[index]:
+            octets[index] = '0'
+    result = ".".join(octets)
```
Comment (Stranger6667) [inline django/utils/ipv4.py]: It can be returned immediately without `result` variable

File: django/utils/ipv4.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+def clean_ipv4_address(ip_str):
+    """
+    Cleans an IPv4 address string.
+    Args:
+        ip_str: A valid IPv4 address.
+
+    Returns:
+        A normalized IPv4 address, or the same value
+    """
+    octets = ip_str.split(".")
```
Comment (frankvdp) [inline django/utils/ipv4.py]: You _could_ turn this into a oneliner ...
```
'.'.join([str(int(num)) for num in string.split('.')])
```

**Comment by timgraham** [comment]
Hi, any chance to finish this up soon or should we close the pull request for now?

**Comment by darjeeling** [comment]
@timgraham, sorry for lateness.
I think I can update in next 3 days.
If I can't update, close at that time thanks.

**Comment by timgraham** [comment]
Feel free to send a new PR if you have time to return to it.

**Pull request closed without merging**